src/rabbit_dead_letter.erl

```diff
@@ -84,36 +84,90 @@ x_death_event_key(Info, Key, KeyType) ->
         {value, {Key, KeyType, Val}} -> Val
     end.
 
+maybe_append_to_event_group(Table, _Key, _SeenKeys, []) ->
+    [Table];
+maybe_append_to_event_group(Table, {_Queue, _Reason} = Key, SeenKeys, Acc) ->
+    case sets:is_element(Key, SeenKeys) of
+        true  -> Acc;
+        false -> [Table | Acc]
+    end.
+
+group_by_queue_and_reason([]) ->
+    [];
+group_by_queue_and_reason([Table]) ->
+    [Table];
+group_by_queue_and_reason(Tables) ->
+    {_, Grouped} =
+        lists:foldl(fun ({table, Info}, {SeenKeys, Acc}) ->
+            Q = x_death_event_key(Info, <<"queue">>, longstr),
+            R = x_death_event_key(Info, <<"reason">>, longstr),
+            Matcher = queue_and_reason_matcher(Q, R),
+            {Matches, _} = lists:partition(Matcher, Tables),
+            {Augmented, N} = case Matches of
+                                 [X]        -> {X, 1};
+                                 [X|_] = Xs -> {X, length(Xs)}
+                             end,
+            Key = {Q, R},
+            Acc1 = maybe_append_to_event_group(
+                ensure_xdeath_event_count(Augmented, N), Key, SeenKeys, Acc),
+            {sets:add_element(Key, SeenKeys), Acc1}
+        end, {sets:new(), []}, Tables),
```

Comment:
perhaps since you use a set, why not use the set as the actual accumulator, and then do http://erlang.org/doc/man/sets.html#to_list-1

This is more of a _taste_ comment than a requirement.